sql-plugin/src/main/scala/com/nvidia/spark/rapids/SplitColumnarBatchProcessor.scala

```diff
@@ -0,0 +1,333 @@
+   *  [ (7, [abc, bc]), (10, [aa, cc]), (11, [ade, fd]) ]
+   * ]
+   *
+   * @return An array of batches, containing (row counts, partition values) pairs,
+   *         such that each batch's size is less than cudfLimit.
+   */
+  private def splitPartitionDataIntoGroups(partRows: Array[Long], partValues: Array[InternalRow],
+      partSchema: StructType): Array[Array[(Long, InternalRow)]] = {
+    val resultBatches = ArrayBuffer[Array[(Long, InternalRow)]]()
+    val currentBatch = ArrayBuffer[(Long, InternalRow)]()
+    // Initialize a HashMap to store the size of the current batch for each column
+    val sizeOfBatch = mutable.HashMap.empty[Int, Long]
+    var partIndex = 0
+    while (partIndex < partRows.length) {
+      val rowsInPartition = partRows(partIndex)
+      val valuesInPartition = partValues(partIndex)
+      // Add the partition's data to the current batch
+      currentBatch.append((rowsInPartition, valuesInPartition))
+      partSchema.zipWithIndex.foreach {
+        case (field, colIndex) if field.dataType == StringType =>
+          val singleValue = valuesInPartition.getString(colIndex)
+          val sizeOfSingleValue = singleValue.getBytes("UTF-8").length.toLong
+          val sizeOfPartition = rowsInPartition * sizeOfSingleValue
+          // Check if adding the partition to the current batch exceeds cuDFLimit
+          if (sizeOfBatch.getOrElseUpdate(colIndex, 0L) + sizeOfPartition <= cuDF_LIMIT) {
+            sizeOfBatch(colIndex) += sizeOfPartition
+          } else {
+            // Need to split the current partition
+            val newRows = (cuDF_LIMIT - sizeOfBatch(colIndex)) / sizeOfSingleValue
+            // Update batches and clear states
+            currentBatch.update(currentBatch.length - 1, (newRows, valuesInPartition))
+            resultBatches.append(currentBatch.toArray)
+            currentBatch.clear()
+            sizeOfBatch.clear()
+            // Reprocess the current partition with remaining rows
+            partRows(partIndex) -= newRows
+            partIndex -= 1
+          }
+        case _ =>
+        // Assumption: Columns of other data type would not have fit already. Do nothing.
+      }
+      partIndex += 1
+    }
+    // Add the last remaining batch to resultBatches
+    if (currentBatch.nonEmpty) {
+      resultBatches.append(currentBatch.toArray)
+    }
+    resultBatches.toArray
+  }
+
+  /**
+   * Splits the input ColumnarBatch into smaller batches, wraps these batches with partition
+   * data, and returns an Iterator.
+   *
+   * Note:
+   * - Partition values are not converted to GPU ColumnVectors and merged with the smaller
+   *   ColumnarBatch to avoid wasting GPU memory.
+   * - The actual creation of GPU ColumnVectors and merging is handled by the
+   *   SplitBatchIteratorWithPartitionData lazily.
+   *
+   * @param batch           Input ColumnarBatch.
+   * @param splitRowData    Array of arrays containing split row data (RowNums and InternalRows).
```

Comment:
Nit: Would be good to elaborate a bit on how this should be interpreted, e.g.: what are the expected semantics when moving between top-level arrays?